### DISCUSSION TITLE
feat: VRMの非破壊編集に対応する

### DISCUSSION THREAD

**Discussion by kaikoga** [anatawa12/AvatarOptimizer#522] [Ideas]
## 欲しいもの

ndmf をVRChat以外のアバター形式に対応させる (Discussions: https://github.com/bdunderscore/ndmf/discussions/39 ) に伴い、 VRM0 / VRM1のアバターを非破壊編集で作れるように管理したいと考えています。

Avatar Optimizierを用いてVRM0 / VRM1を作る際、現状ではManual Bake Avatar後に手動でVRMセットアップが必要であり、毎回手作業で行うのは大変なので、Avatar OptimizierがVRMのコンポーネントを扱えるようにしたいです。

as-is:

- Manual Bake Avatarする
- 手動でVRMのコンポーネントをセットアップする ← ここが破壊的編集
- Export to VRM0 / VRM1 する

to-be:

- VRMのコンポーネントをセットアップしたプレハブを用意する
- Manual Bake Avatarする
- そのまま Export to VRM0 / VRM1 する

現在はVRCSDKとUniVRMは共存できませんが、近いうちに共存できるようになるため、VRM対応のAvatar Optimizierのforkではなく、本体へのVRM対応の組み込みを目指したいと考えています。

## 努力目標

- UniVRMが動作するいずれかのバージョンのUnityに対応する
  - Unity 2021で動くと嬉しいけど、必須ではない
- VRCSDKとUniVRMが共存できない場合に備え、UPMに対応する（VPMパッケージはそのまま使えるはず）
- VRM0系 / VRM1系の両方に対応する
  - どちらかを削って楽になるタイプの作業ではないと思うし、作業者は自分になると思うので・・・

**Comment by kaikoga** [comment]
## タスク

kaikoga/AvatarOptimizer#1 にある内容を本体に対してやる（このPR自体はPoCであり、また随時破壊されるので直接マージはしない）

- [ ] 必要な場所で`#if UNITY_2020_2_OR_NEWER` する
- [ ] "com.vrchat.avatars" に対して Version Define `AAO_VRCSDK3_AVATARS` を追加する
- [ ] アバター本体への参照として VRCAvatarDescriptor を使っているところを可能な限り外す
- [ ] VRCSDKがないと動かないコンポーネントを `AAO_VRCSDK3_AVATARS` で無効化する
- [ ] VRCSDKのコンポーネントへの参照を `AAO_VRCSDK3_AVATARS` に押し込める
- [ ] 自動生成コードが条件コンパイルに対応していないところをどうにかする

**Comment by kaikoga** [comment]
## 議論の余地があること

#### VRCSDKのデータ型をシリアライズしてるコンポーネントの扱い

- 例： AAO Merge PhysBone みたいなやつ
- VRCSDKがない環境でプレハブを読み込むとシリアライズされたデータが破壊されるけど良い？

**Comment by kaikoga** [comment]
## 議論の余地があること

#### VRCSDKがないと動かないコンポーネントの表示

- 例： AAO Clear Endpoint Position とか
- kaikoga/AvatarOptimizer#1 でやってるような感じで良いと思っています
  - 編集できるのであれば、編集を禁止するまででもないと思う（今回の例だと設定項目は存在しないけど）

**Reply by anatawa12** [reply]
私の認識ではそもそもMergePhysBoneやClearEndpointPosition等はVRCSDKない環境でコンポーネント全体を`#if`でけしちゃっていいと思ってますけど消さない理由はどんな感じでしょうか？

**Reply by kaikoga** [reply]
VRCSDKのある環境で作成した衣装プレハブをVRCSDKのない環境へ持って行った場合にMissingになるのを避けたいという考えがありました。
例えば、衣装製作者がVRCSDKのある環境を想定した衣装プレハブのみを配布していて、VRCSDKのない環境でその衣装を読み込んだ場合に、Missingとして対処するよりは少なくとも何のコンポーネントが刺さっていたかがわかった方がユーザー側で対処がしやすいのではないかと思っています。
また、コンポーネントに対して処理を行うPassを丸ごと `#if` で消してしまうことで、「VRCSDKのない環境では何もしないコンポーネント」にすることができます。

コンポーネントの存在ごと消し去るでも大きな問題はないと思います。

**Reply by kaikoga** [reply]
https://github.com/bdunderscore/modular-avatar/discussions/467#discussioncomment-7165254 でも書いたのですが、「Missingなコンポーネントが刺さっているPrefabのVariantを作れない」は渋いかなと思います。

**Reply by anatawa12** [reply]
Variant層でRemove Componentじゃだめなんですっけ

**Reply by kaikoga** [reply]
改めて自分が確認したところの挙動としては、こんな感じになりました。

- VRCSDKがある環境でVRCSDKのコンポーネントを剥がしたPrefab VariantをVRCSDKがない環境に持っていくことは可能で、普通に動く
  - この状態のPrefab VariantをVRCSDKがない環境で複製することはできる
- （VRCSDKがない環境で）**Missingが刺さっているPrefabのVariantを作ろうとするとUnityに阻止される**
- Missingが刺さっているPrefabをシーンに置くことはできる
  - **この状態からMissingを剥がしたものをPrefab Variantとして保存することはできる**

まとめると、Missingにされる側からしてみれば多少不便な程度で済みますが、Missingにしなくて済むものをわざわざMissingにする理由としては薄いです。

**Reply by anatawa12** [reply]
Variant作るのにMissingを消す必要があるのであれば、MergePhysBone, ClearEndpointPositionが存在する時点でPhysBoneのコンポーネントがすでにMissingで存在しているのでMissingにせず残しておくメリットが、意味のないコンポーネントが追加できたりRequireComponent周りでifdefがふえるデメリットに対して大きいとは私は思いません。

VRCSDKがないと意味がないけどVRCSDKのコンポーネントに依存してない場合には残すのは悪くないと思います

**Reply by kaikoga** [reply]
一旦Missingで進めてあとから何もしないダミーのコンポーネントを追加する方法を具体的なPull Requestで考える進め方はできる話なので、それで良いですか？

**Reply by anatawa12** [reply]
はい

**Comment by anatawa12** [comment]
VRM対応については私は入れるのには前向きですが、完全にコミュニティによるメンテナンスの形で有ることを条件にしたいと考えてます。
また、UniVRM等の対応はversion definesすることになると思うため、UPM Package版のみの対応になりそうだなと考えています。

**Reply by kaikoga** [reply]
UniVRMをunitypackageでインストールした場合はAvatarOptimizerのVRM対応コードが動かないという点は、ドキュメントか警告表示で対応する必要がありそうですね。

**Comment by kaikoga** [comment]
こちらも「VRCSDKへの依存を取り除く」と「（Uni）VRM特有の対応を追加する」は工程として分割可能なので、 [Modular Avatar と同様に](https://github.com/bdunderscore/modular-avatar/issues/478) まずはVRCSDKへの依存を取り除くタスクをIssue化しようと考えています。

more-ndmf branchの差分が大きそうなので、more-ndmfの取り込みを待って作業を開始しようと思います。

**Reply by kaikoga** [reply]
VRCSDKへの依存の解消を始めます https://github.com/anatawa12/AvatarOptimizer/issues/607

**Reply by kaikoga** [reply]
VRM特有の対応を追加するIssue #651
